base/boot.jl

```diff
@@ -562,7 +562,24 @@ end
 
 NamedTuple() = NamedTuple{(),Tuple{}}(())
 
-NamedTuple{names}(args::Tuple) where {names} = NamedTuple{names,typeof(args)}(args)
+function safe_Typeof(@nospecialize(x))
+    if x isa DataType
+        if x.layout === Ptr{Nothing}(0)  # we don't have C_NULL yet
```

Comment:
this looks.....scary? do we not have an official query function for this?

Reply:
I guess not? `.layout == C_NULL` seems to be the common way in `Base`:

```console
$ rg '\.layout'
reflection.jl
336:    dt.layout == C_NULL && throw(UndefRefError())
337:    alignment = unsafe_load(convert(Ptr{DataTypeLayout}, dt.layout)).alignment
370:    dt.layout == C_NULL && throw(UndefRefError())
371:    flags = unsafe_load(convert(Ptr{DataTypeLayout}, dt.layout)).flags
383:    dt.layout == C_NULL && throw(UndefRefError())
384:    npointers = unsafe_load(convert(Ptr{DataTypeLayout}, dt.layout)).npointers
392:located at `(dt.layout + sizeof(DataTypeLayout))`.
399:    dt.layout == C_NULL && throw(UndefRefError())
400:    flags = unsafe_load(convert(Ptr{DataTypeLayout}, dt.layout)).flags
```

Reply:
I'm afraid of the performance implications of this PR, but why not just use `Typeof` here?

Reply:
By performance, do you mean compile-time?

I'm checking `.layout` as otherwise I get:

```julia
julia> NamedTuple{(:a,),Tuple{Type{Vector.body}}}((Vector.body,))
ERROR: UndefVarError: T not defined
```